Allow cross-group subresource registrations in the APIInstaller.

madhusudancs opened the pull request:
@lavalamp @bgrant0607 @deads2k @smarterclayton @liggitt @piosz 

This is a rework of #20745 and #19802 based on all your reviews. Only the last commit needs review here. The other commits are being reviewed separately in #21603.

madhusudancs committed:
Allow cross-group subresource registrations in the APIInstaller.

This allows subresources which belong to different API groups than
their parents to be registered in the APIInstaller and REST handlers
installed for them. The specific changes that makes this possible
are:

1. Allow subresource overrides to be specified while registering an
   API group.
2. Use those overrides in the APIInstaller while validating the
   resource/subresource group version to allow subresources which
   belong to a different group to be registered if they have an
   override specified.
3. Use the RESTMapper supplied in the override to map the REST paths
   to the correct subresource storage object, i.e. correct group
   version kinds.

deads2k commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
 	fqKinds, err := a.group.Typer.ObjectKinds(object)
 	if err != nil {
-		return nil, err
+		return fqKindToRegister, err
```
This is just an empty `unversioned.GroupVersionKind`?  Please return that explictly to make it clear the value is meant to be invalid/unused.

madhusudancs replied:
Done.

deads2k commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
+	fqKindToRegister := unversioned.GroupVersionKind{}
+	allowedGroup := a.group.GroupVersion.Group
+	if override, ok := a.group.SubresourceOverrides[path]; ok {
+		allowedGroup = override.GroupVersion.Group
```
So the `SubresourceOverrides` explicitly provides a `GroupVersion`, but we're ignoring the `version` part of that information?  Why?

deads2k replied:
Yeah, since version is definitely expected in the return and you're using this field as the marker, why is only half the field respected?  It almost seems like what you want is to take the `GroupVersion` from here and do a `.WithKind` from the `ObjectKinds`, which matches the expected group.

bgrant0607 replied:
I had the same question.

madhusudancs replied:
You are right and I realize this thing is wrong. I should take both group and version from the override. I was trying hard to keep the old code as much as possible. I have fixed this.

deads2k commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
+		// DEBUG: if fqKind.Group == a.group.GroupVersion.Group
+		if fqKind.Group == allowedGroup {
+			fqKindToRegister.Group = fqKind.Group
+			fqKindToRegister.Version = a.group.GroupVersion.Version
```
This doesn't look right.  Why would I take group information from one source and smash it together with version information from another source which is referring to a potentially unrelated `GroupVersion`?  I could probably see it if `allowedGroup == a.group.GroupVersion.Group` (in which case this entire loop in unnecessary), but why do this otherwise?

bgrant0607 replied:
+1

madhusudancs replied:
Fixed this, same as above.

deads2k commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,62 +110,54 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
 		}
 
 		// TODO This keeps it doing what it was doing before, but it doesn't feel right.
 		if fqKind.Group == extensions.GroupName && fqKind.Kind == "ThirdPartyResourceData" {
```
Do you know why this is special-cased?  I still don't, but this pull is changing behavior around this bit of hardcoding (unlike the previous comment) so I think it should be re-visited.

bgrant0607 replied:
Let's deal with ThirdPartyResource in another PR. The currently implemented behavior isn't what we want.

madhusudancs replied:
Yeah, I think that should be in another PR.

deads2k commented on pkg/apiserver/api_installer.go:
```diff
@@ -948,3 +973,19 @@ var _ rest.StorageMetadata = defaultStorageMetadata{}
+		// TODO: support deeper paths
+		return "", "", false, fmt.Errorf("api_installer allows only one or two segment paths (resource or resource/subresource)")
+	}
+	return resource, subresource, len(subresource) > 0, nil
```
This bool doesn't seem to add much value.  Why not return a single field that can't possibly disagree with itself?

madhusudancs commented:
@lavalamp I have addressed your comments, PTAL.

@bgrant0607 @deads2k PTAL too.

lavalamp commented:
Sorry, I didn't get all the way through this yesterday, so I may have more comments. Getting through this is my priority for today.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,62 +111,55 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
+// storage object. If the storage object is a subresource, it just return's the object's group
+// version kind without validating it's group against the group we are installing the REST
+// handlers for here.
+func (a *APIInstaller) validFullyQualifiedKind(path string, storage rest.Storage, hasSubresource bool) (unversioned.GroupVersionKind, error) {
```
Suggest: getBodyKind, getWireKind, getResourceKind

The validation isn't important to highlight in the function name. The thing to call attention to is that this produces the kind corresponding to the return type, not the kind corresponding to the path the resource is mounted under.

madhusudancs replied:
Done.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -187,20 +181,54 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		parentFQKindToRegister := unversioned.GroupVersionKind{}
```
Can you replace this section with a call to the other function?

madhusudancs replied:
Done.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -187,20 +181,54 @@ func (a *APIInstaller) registerResourceHandlers(path string, storage rest.Storag
 		if err != nil {
 			return nil, err
 		}
 		mapping.Scope = parentMapping.Scope
```
Actually, can you just re-call this current function?

madhusudancs replied:
Refactored this a bit. So this should be fixed.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,62 +111,55 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
+	mapper := a.group.Mapper
+
+	if hasSubresource {
+		groupMeta, err := registered.Group(fqKindToRegister.Group)
```
Reading this function, I cannot tell if Job's scale subresource will end up exposed at `apis/batch/v1/job/scale` or `apis/autoscaling/v1/job/scale`. The former is correct. Can you add comments until it is clear?

madhusudancs replied:
Job will not have a Scale.

lavalamp replied:
Sorry, I keep forgetting. Pretend I said replicaset or deployment.

On Fri, Feb 26, 2016 at 2:17 PM, Madhusudan.C.S <notifications@github.com>
wrote:

> In pkg/apiserver/api_installer.go
> <https://github.com/kubernetes/kubernetes/pull/21909#discussion_r54309928>
> :
>
> >
> > -	versionedPtr, err := a.group.Creater.New(a.group.GroupVersion.WithKind(kind))
> > -	if err != nil {
> > -		return nil, err
> > +// restMapping returns the RESTMapping object for the given resource or the subresource based on
> > +// the supplied subresource overrides.
> > +func (a *APIInstaller) restMapping(path, resource string, hasSubresource bool, fqKindToRegister unversioned.GroupVersionKind) (*meta.RESTMapping, error) {
> > +	mapper := a.group.Mapper
> > +
> > +	if hasSubresource {
> > +		groupMeta, err := registered.Group(fqKindToRegister.Group)
>
> Job will not have a Scale.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21909/files#r54309928>.
>

madhusudancs replied:
Fixed with examples. I think that's the best way to make this clear.

lavalamp commented:
This needs a test and then I think it's OK.

lavalamp commented on pkg/apiserver/apiserver.go:
```diff
@@ -96,6 +96,11 @@ type APIGroupVersion struct {
 
 	MinRequestTimeout time.Duration
+
+	// SubresourceOverrides contains the GroupVersionKind overrides for each subresource that is
```
Comment doesn't match any more. Mention "external version" somewhere.

madhusudancs replied:
Fixed.

lavalamp commented on pkg/genericapiserver/genericapiserver.go:
```diff
@@ -192,6 +192,11 @@ type APIGroupInfo struct {
 	// ParameterCodec performs conversions for query parameters passed to API calls
 	ParameterCodec runtime.ParameterCodec
+
+	// SubresourceOverrides contains the GroupVersionKind overrides for each subresource that is
```
Same comment.

madhusudancs replied:
Fixed.

lavalamp commented on pkg/apiserver/apiserver.go:
```diff
@@ -96,6 +96,11 @@ type APIGroupVersion struct {
+	// SubresourceOverrides contains the GroupVersionKind overrides for each subresource that is
+	// accessible from this API group version. The key of this map should be the path of the
+	// subresource. The keys here should match the keys in the Storage map above for subresources.
+	SubresourceGroupVersionKind map[string]unversioned.GroupVersionKind
```
Why did you need to add it both here and in genericapaiserver? Should only need to be added in one place-- genericapiserver is preferable.

madhusudancs replied:
As we discussed offline, this is required to plumb the value from master to api_installer.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,96 +110,90 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
 
-	kind := fqKindToRegister.Kind
+// restMapping returns the RESTMapping object for the given resource or the subresource based on
+// the supplied subresource overrides.
```
Fix comment. "returns rest mapper for the resource, or the resource's parent if it is a sub resource."

madhusudancs replied:
Done.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,96 +110,90 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
-	kind := fqKindToRegister.Kind
+// restMapping returns the RESTMapping object for the given resource or the subresource based on
+// the supplied subresource overrides.
+func (a *APIInstaller) restMapping(path, resource string, hasSubresource bool, fqKindToRegister unversioned.GroupVersionKind) (*meta.RESTMapping, error) {
```
It's extremely strange that this function expects fqKindToRegister to not correspond to resource.

Like maybe just don't take path, hasSubresource or fqKindToRegister as parameters and always do the section in the `if`.

madhusudancs replied:
Done.

lavalamp commented:
OK, this is looking like it might plausibly work now :) :)

...but I'm still going to ask for a test.

madhusudancs commented:
> ...but I'm still going to ask for a test.

Figuring out how to write a test for this :) Will push the test as soon as I have it working.

madhusudancs commented:
@lavalamp Added a test and addressed other comments. I have created new commits (last two) for all the changes since last review to make the reviews easier. PTAL.

madhusudancs commented:
@k8s-bot test this issue: #22088

lavalamp commented on pkg/apiserver/apiserver_test.go:
```diff
@@ -144,6 +152,11 @@ func addTestTypes() {
 		&apiservertesting.SimpleGetOptions{}, &apiservertesting.SimpleRoot{})
+	api.Scheme.AddKnownTypes(testGroup2Version,
+		&SimpleXGSubresource{})
+	api.Scheme.AddKnownTypes(testGroupVersion, &api.Pod{})
```
Is registering pod necessary?

madhusudancs replied:
Removed.

Sorry, that was oversight on my part.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,97 +110,94 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
+//      Resource:    extensions/v1beta1.ReplicaSet
+//      Subresource: autoscaling/v1.Scale
+//      REST path:   /apis/extensions/v1beta1/namespaces/{namespace}/replicaset/{name}/scale
+func (a *APIInstaller) restMapping(resource string) (*meta.RESTMapping, error) {
```
'resource' should never be a subresource, correct? Please fix the comments :)

madhusudancs replied:
Done.

I removed ", or the resource's parent if it is a sub resource" part from the first line. But I am leaving the subresource example as is. I think that adds some clarity about how the subresource paths are mapped.

lavalamp replied:
I agree the example is helpful.

liggitt replied:
can you clarify the comment... it looks like you could pass in "batch/v1.Job" as the `resource` arg

lavalamp replied:
+1, clarify what resource accepts.

madhusudancs replied:
Done.

lavalamp commented on pkg/apiserver/apiserver_test.go:
```diff
@@ -3157,6 +3170,106 @@ func TestUpdateChecksAPIVersion(t *testing.T) {
+
+	if itemOut.Name != subresourceStorage.item.Name {
+		t.Errorf("Unexpected data: %#v, expected %#v (%s)", itemOut, subresourceStorage.item, string(body))
+	}
```
Test that the returned item is the correct group/version/kind.

madhusudancs replied:
Done.

lavalamp commented on pkg/apiserver/apiserver_test.go:
```diff
@@ -144,6 +152,11 @@ func addTestTypes() {
+		&SimpleXGSubresource{})
+	api.Scheme.AddKnownTypes(testGroupVersion, &api.Pod{})
+	api.Scheme.AddKnownTypes(testInternalGroup2Version,
+		&SimpleXGSubresource{})
```
Maybe register this in the other group, too, to prove that the correct one is chosen.

madhusudancs replied:
Done.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -110,97 +110,94 @@ func (a *APIInstaller) NewWebService() *restful.WebService {
+}
 
-	kind := fqKindToRegister.Kind
+// restMapping returns rest mapper for the resource, or the resource's parent if it is a sub resource.
```
Remove second half of sentence and replace with something saying not to pass a subresource to this function.

lavalamp replied:
Oh, I guess you haven't pushed yet :) sorry!

liggitt commented:
do we actually want to add "replicationcontrollers/scale" in v1 to prove this works?

deads2k commented:
> do we actually want to add "replicationcontrollers/scale" in v1 to prove this works?

Great idea.

lavalamp commented:
That may be a good idea.

On Tue, Mar 1, 2016 at 12:38 PM, David Eads <notifications@github.com>
wrote:

> do we actually want to add "replicationcontrollers/scale" in v1 to prove
> this works?
>
> Great idea.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21909#issuecomment-190890930>
> .
>

madhusudancs commented:
@liggitt @deads2k 

> do we actually want to add "replicationcontrollers/scale" in v1 to prove this works? I assume we would make it use the autoscaling/v1 Scale kind.

Yes. That was how I started with this whole cross group registration thing, that's how I ran into all sorts of problems that I am fixing right now ;-) There is a PR for that already - #21910

madhusudancs commented:
@lavalamp I have pushed the changes. PTAL.

lavalamp commented on pkg/apiserver/apiserver_test.go:
```diff
@@ -3157,6 +3172,121 @@ func TestUpdateChecksAPIVersion(t *testing.T) {
+	// to output type object. So it's values don't appear in the decoded output object.
+	decoder := json.NewDecoder(strings.NewReader(body))
+	var itemFromBody SimpleXGSubresource
+	decoder.Decode(&itemFromBody)
```
check error return!

madhusudancs replied:
Done.

lavalamp commented:
Two nits, then LGTM

liggitt commented on pkg/apiserver/apiserver_test.go:
```diff
@@ -3157,6 +3172,121 @@ func TestUpdateChecksAPIVersion(t *testing.T) {
+	// installed decoders. TypeMeta cannot be decoded because it is added to the ignored
+	// conversion type list in API scheme and hence cannot be converted from input type object
+	// to output type object. So it's values don't appear in the decoded output object.
+	decoder := json.NewDecoder(strings.NewReader(body))
```
This doesn't seem right… we can't decode an API object using a real codec?

madhusudancs replied:
@liggitt We can decode a serialized API object into a destination, not just `TypeMeta`. In fact we can decode `TypeMeta` just fine, but the decoded values won't be copied to the destination object (`itemOut` here) and that is by design. It would be dangerous to overwrite `TypeMeta`. We could probably implement a converter to allow overwriting `TypeMeta` in tests, but I don't think it is worth spending time on it.

I think this is a reasonable workaround. Do you have any other alternative suggestions?

lavalamp replied:
@liggett it's that I asked for the version/kind to be checked, and our codec consumes that information and leaves those fields blank.

liggitt replied:
UnstructuredJSONScheme would do this, I think...
```
obj, gvk, err := runtime.UnstructuredJSONScheme.Decode(data, nil, nil)
// assert no err
// assert gvk matches
```

liggitt replied:
I guess it's json decoding either way, I just thought there was an existing codepath we could use

madhusudancs replied:
@liggitt Do you feel strongly about swapping the current unmarshaling code with this snippet?

liggitt replied:
I guess not

madhusudancs replied:
@liggitt Alright then, thanks!

madhusudancs commented:
@lavalamp Addressed the comments. PTAL.

lavalamp commented:
LGTM

lavalamp commented:
Squash please, then apply the label :)
